### PR TITLE
Support $ and / in ccr follow

### LINKED ISSUE #99078: It is impossible to create a follower index with `$` character in its name
### Description

It is impossible to create a follower index with a `$` character in its name (please note, it is possible to create a regular index with `$` in its name).

This could be reproduced by rest request:
```
PUT /my-index/_ccr/follow
{
    "remote_cluster": "my-remote-cluster",
    "leader_index": "my-index$follower"
}
```

or `CcrIntegTestCase`:

```
    public void testFollowIndex() throws Exception {
        assertAcked(leaderClient().admin().indices().prepareCreate("my$index"));
        setupRemoteCluster();
        final PutFollowAction.Request followRequest = putFollow("my$index", "my$index");
        followerClient().execute(PutFollowAction.INSTANCE, followRequest).get();
    }
```

This is happening as during initial follow phase we restore the leader cluster from the remote using a snapshot restore.
Corresponding request is created with index name added to `RestoreSnapshotRequest#renamePattern` (regexp) without escaping `$`.

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Currently the renameReplacement variable in the restore snapshot request
is intended to be a string literal for ccr. However, when a string is passed
to replaceAll in java $ and / will be treated as special characters. This
means that following indices with $ in them breaks. This commit fixes the
issue by quoting the special characters before calling replaceAll.

Fixes https://github.com/elastic/elasticsearch/issues/99078.

### CODE REVIEW

File: x-pack/plugin/ccr/src/internalClusterTest/java/org/elasticsearch/xpack/ccr/IndexFollowingIT.java [hunk 1]
```diff
@@ -630,6 +630,12 @@ public void testFollowNonExistentIndex() throws Exception {
         );
     }
 
+    public void testFollowWith$InIndexName() throws Exception {
+        String indexSettings = getIndexSettings(1, 0);
+        assertAcked(leaderClient().admin().indices().prepareCreate("test-leader$").setSource(indexSettings, XContentType.JSON).get());
+        followerClient().execute(PutFollowAction.INSTANCE, putFollow("test-leader$", "test-leader$")).actionGet();
```
Comment (idegtiarenko) [inline x-pack/plugin/ccr/src/internalClusterTest/java/org/elasticsearch/xpack/ccr/IndexFollowingIT.java:636]:
```suggestion
        followerClient().execute(PutFollowAction.INSTANCE, putFollow("test-leader$", "follower-leader$")).actionGet();
```

Should we also check that the follower index actually started?

File: server/src/main/java/org/elasticsearch/snapshots/RestoreService.java [hunk 2]
```diff
@@ -693,7 +694,7 @@ private Set<Index> resolveSystemIndicesToDelete(ClusterState currentState, Set<S
     static DataStream updateDataStream(DataStream dataStream, Metadata.Builder metadata, RestoreSnapshotRequest request) {
         String dataStreamName = dataStream.getName();
         if (request.renamePattern() != null && request.renameReplacement() != null) {
-            dataStreamName = dataStreamName.replaceAll(request.renamePattern(), request.renameReplacement());
+            dataStreamName = dataStreamName.replaceAll(request.renamePattern(), Matcher.quoteReplacement(request.renameReplacement()));
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/snapshots/RestoreService.java:697]: Is changing it on this level going to break snapshot restore with a replacement pattern?

I wonder if we should instead quote it at:
https://github.com/elastic/elasticsearch/blob/f4e3113ad00541f084b69664c6476c8fcf737386/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java#L188

I do not think we ever accept follower index in a put follower api as a regexp and that way we are not going to change existing snapshot restore logic.

Comment (Tim-Brooks) [reply server/src/main/java/org/elasticsearch/snapshots/RestoreService.java]: When I was looking at this last week, I thought it was a bug that all name replacements were treated as regex. I assumed that those were supposed to be string literals. But now I see that we support match groups for normal restores, so I changed this to only impact CCR.

**Review by idegtiarenko — APPROVED** [review]
I think the test failure is unrelated to the change. :+1:

**Comment by arteam** [comment]
I believe the failed test has already been muted in https://github.com/elastic/elasticsearch/pull/100192

**Review by arteam — APPROVED** [review]
LGTM2

**Pull request merged**